File: angular/src/app/service/settings.service.ts
import {HttpClient} from '@angular/common/http';
import {lastValueFrom, Observable, Subject} from 'rxjs';
import {Injectable} from '@angular/core';
import {Settings, SettingsResponse} from '../interface/Settings';
import {environment} from '../../environments/environment';
import {map} from 'rxjs/operators';
import {RestResponse} from '../interface/RestResponse';
import {BaseHttpService} from './basehttp.service';

@Injectable({providedIn: 'root'})
export class SettingsService extends BaseHttpService {
  private cache: Settings;
  private getSettingsPromise: Promise<Settings> = null
  $settings: Subject<Settings>

  constructor(private http: HttpClient) {
    super();
    this.$settings = new Subject<Settings>();
  }

  getSettings(): Promise<Settings> {
    if (this.cache) {
      return Promise.resolve(this.cache);
    }

    if (this.getSettingsPromise == null) {
      this.getSettingsInternal()
    }

    return this.getSettingsPromise
  }

  private getSettingsInternal() {
    this.getSettingsPromise = lastValueFrom(this.http.get<SettingsResponse>(`${environment.api}/settings/`).pipe(
      map(res => {
        if (res.success) {
          const settings = res.data as Settings;

          // order
          settings.productAttributeOrderMap = new Map<string, number>()
          settings.categoryAttributeOrderMap = new Map<string, number>()
          settings.pimLocaleOrderMap = new Map<string, number>()
          for (const [i, attr] of settings.productAttributes.entries()) {
            attr.order = i
            settings.productAttributeOrderMap.set(attr.name, attr.order)
          }
          for (const [i, attr] of settings.categoryAttributes.entries()) {
            attr.order = i
            settings.categoryAttributeOrderMap.set(attr.name, attr.order)
          }
          for (const [i, loc] of settings.pimLocales.entries()) {
            loc.order = i
            settings.pimLocaleOrderMap.set(loc.languageCode, loc.order)
          }

          this.cache = settings
          this.$settings.next(settings)
          return settings;
        } else {
          throw Error(res.data as string)
        }
      })
    ))
  }

  async reloadSettings(): Promise<Settings> {
    this.cache = null;
    return await this.getSettings();
  }

  addPimLocale(name: string, countryCode: string, languageCode: String): Observable<string> {
    return this.http.post<RestResponse<string>>(`${environment.api}/settings/pimLocale`, {
      name: name,
      countryCode: countryCode,
      languageCode: languageCode
    }).pipe(
      map(this.handleResponse)
    )
  }

  addCategoryAttribute(name: string, description: string, localizable: boolean, valueType: String): Observable<string> {
    return this.http.post<RestResponse<string>>(`${environment.api}/settings/categoryAttribute`, {
      name: name,
      description: description,
      localizable: localizable,
      valueType: valueType
    }).pipe(
      map(this.handleResponse)
    )
  }

  addProductAttribute(name: string, description: string, localizable: boolean, valueType: String): Observable<string> {
    return this.http.post<RestResponse<string>>(`${environment.api}/settings/productAttribute`, {
      name: name,
      description: description,
      localizable: localizable,
      valueType: valueType,
      variation: true,
      options: [] // TODO
    }).pipe(
      map(this.handleResponse)
    )
  }

  deletePimLocale(id: number): Observable<string> {
    return this.http.request<RestResponse<string>>(
      'delete', `${environment.api}/settings/pimLocale`, this.buildDeleteRequestOption(id)).pipe(
      map(this.handleResponse)
    )
  }

  deleteCategoryAttribute(id: number): Observable<string> {
    return this.http.request<RestResponse<string>>(
      'delete', `${environment.api}/settings/categoryAttribute`, this.buildDeleteRequestOption(id)).pipe(
      map(this.handleResponse)
    )
  }

  deleteProductAttribute(id: number): Observable<string> {
    return this.http.request<RestResponse<string>>(
      'delete', `${environment.api}/settings/productAttribute`, this.buildDeleteRequestOption(id)).pipe(
      map(this.handleResponse)
    )
  }

  private handleResponse = res => {
    if (res.success) {
      return res.data
    } else {
      throw Error(res.data)
    }
  }
}

File: back/src/main/kotlin/com/catprogrammer/pim/service/SettingsService.kt
package com.catprogrammer.pim.service

import com.catprogrammer.pim.dto.SettingsDto
import com.catprogrammer.pim.entity.CategoryAttribute
import com.catprogrammer.pim.entity.PimLocale
import com.catprogrammer.pim.entity.ProductAttribute
import com.catprogrammer.pim.enumeration.AttributeValueType
import com.catprogrammer.pim.repository.CategoryAttributeRepository
import com.catprogrammer.pim.repository.PimLocaleRepository
import com.catprogrammer.pim.repository.ProductAttributeRepository
import org.springframework.stereotype.Service

@Service
class SettingsService(
    private val categoryAttributeRepository: CategoryAttributeRepository,
    private val productAttributeRepository: ProductAttributeRepository,
    private val pimLocaleRepository: PimLocaleRepository
) {
    fun getSettings(): SettingsDto = SettingsDto(
        categoryAttributeRepository.findAll(),
        productAttributeRepository.findAll(),
        pimLocaleRepository.findAll()
    )

    fun getPimLocales(): List<PimLocale> = pimLocaleRepository.findAll()

    fun addCategoryAttribute(name: String, localizable: Boolean, valueType: AttributeValueType, description: String) =
        categoryAttributeRepository.save(CategoryAttribute(name, localizable, valueType, description))

    fun addProductAttribute(name: String, localizable: Boolean, valueType: AttributeValueType, variation: Boolean, description: String) =
        productAttributeRepository.save(ProductAttribute(name, localizable, valueType, variation, description))

    fun deleteCategoryAttribute(id: Long) = categoryAttributeRepository.deleteById(id)

    fun deleteProductAttribute(id: Long) = productAttributeRepository.deleteById(id)

    fun addPimLocale(name: String, languageCode: String, countryCode: String) =
        pimLocaleRepository.save(PimLocale(name, languageCode, countryCode))

    fun deletePimLocale(id: Long) = pimLocaleRepository.deleteById(id)
}
